Stratis.Bitcoin.Features.Consensus/ConsensusLoop.cs

```diff
@@ -62,78 +164,188 @@ public void Initialize()
+                {
+                    item.Error = ex.ConsensusError;
+                }
+
+                if (item.Error != null)
+                {
+                    this.logger.LogError("Block rejected: {0}", item.Error.Message);
+
+                    // Pull again.
+                    this.Puller.SetLocation(this.Tip);
+
+                    if (item.Error == ConsensusErrors.BadWitnessNonceSize)
+                    {
+                        this.logger.LogInformation("You probably need witness information, activating witness requirement for peers.");
+                        this.connectionManager.AddDiscoveredNodesRequirement(NodeServices.NODE_WITNESS);
+                        this.Puller.RequestOptions(TransactionOptions.Witness);
+                        return;
+                    }
+
+                    // Set the chain back to ConsensusLoop.Tip.
+                    this.Chain.SetTip(this.Tip);
+                    this.logger.LogTrace("Chain reverted back to block '{0}' ", this.Tip);
+
+                    // Since ChainHeadersBehavior check PoW, MarkBlockInvalid can't be spammed.
+                    this.logger.LogError("Marking block as invalid.");
+                    this.chainState.MarkBlockInvalid(item.Block?.GetHash());
+                }
+                else
+                {
+                    this.logger.LogTrace("Block accepted '{0}' ", this.Tip);
+
+                    this.chainState.HighestValidatedPoW = this.Tip;
+
+                    // We really want to flush if we are at the top of the chain.
+                    // Otherwise, we just allow the flush to happen if it is needed.
+                    bool forceFlush = this.Chain.Tip.HashBlock == item.ChainedBlock?.HashBlock;
+                    this.FlushAsync(forceFlush).GetAwaiter().GetResult();
+
+                    if (this.Tip.ChainWork > this.Chain.Tip.ChainWork)
+                    {
+                        // This is a newly mined block.
+                        this.Chain.SetTip(this.Tip);
+                        this.Puller.SetLocation(this.Tip);
+
+                        this.logger.LogDebug("Block extends consensus tip to '{0}'", this.Tip);
+                    }
+
+                    this.signals.SignalBlock(item.Block);
+                }
             }
 
-            this.logger.LogTrace("(-):*.{0}='{1}',*.{2}='{3}'", nameof(result.ChainedBlock), result.ChainedBlock, nameof(result.Error), result.Error?.Message);
-            return result;
+            this.logger.LogTrace("(-):*.{0}='{1}',*.{2}='{3}'", nameof(item.ChainedBlock), item.ChainedBlock, nameof(item.Error), item.Error?.Message);
         }
 
-        public void AcceptBlock(ContextInformation context)
+        /// <summary>
+        /// Validate a block using the consensus rules.
+        /// </summary>
+        /// <remarks>
+        /// WARNING: This method can only be called from <see cref="ConsensusLoop.AcceptBlock"/>, 
```

Comment:
despite this comment there is PowBlockAssembler.TestBlockValidity which calls this method directly and it seems that it is actually used regularly ...

Reply:
Yeah the next line explains that.
>it the requirement is to validate a block without affecting the consensus db then <see cref="ContextInformation.OnlyCheck"/> must be set to true.

Perhaps breaking this method in two is not a bad idea (one part validates without changing any state) the other updates state (the db, cache)

This will be in another PR but I will add a TODO

Reply:
I see what you mean now, but the comment is really misleading. It should say
`WARNING: This method can only be called from <see cref="ConsensusLoop.AcceptBlock"/>, otherwise <see cref="ContextInformation.OnlyCheck"/> must be set to <c>true</c>, which causes that the method only validates the block without affecting the consensus DB.`